piff/gp_interp_2pcf.py

```
@@ -0,0 +1,372 @@
+
+        self.keys = keys
+        self.npca = npca
+        self.degenerate_points = False
+        self.normalize = normalize
+        self.optimize = optimize
+        self.white_noise = white_noise
+
+        self.kwargs = {
+            'keys': keys,
+            'optimize': optimize,
+            'npca': npca,
+            'kernel': kernel,
+            'normalize':normalize
+        }
+
+        if len(keys)!=2:
+            raise ValueError('the total size of keys can not be something else than 2 using two-point correlation function. Here len(keys) = %i'%(len(keys)))
+
+        if type(kernel) is str:
+            self.kernel_template = [self._eval_kernel(kernel)]
+        else:
+            if type(kernel) is not list and type(kernel) is not np.ndarray:
+                raise TypeError("kernel should be a string a list or a numpy.ndarray of string")
+            else:
+                self.kernel_template = [self._eval_kernel(ker) for ker in kernel]
+
+        self._2pcf = []
+        self._2pcf_dist = []
+        self._2pcf_fit = []
+
+    @staticmethod
+    def _eval_kernel(kernel):
+        # Some import trickery to get all subclasses of sklearn.gaussian_process.kernels.Kernel
+        # into the local namespace without doing "from sklearn.gaussian_process.kernels import *"
+        # and without importing them all manually.
+        def recurse_subclasses(cls):
+            out = []
+            for c in cls.__subclasses__():
+                out.append(c)
+                out.extend(recurse_subclasses(c))
+            return out
+        clses = recurse_subclasses(Kernel)
+        for cls in clses:
+            module = __import__(cls.__module__, globals(), locals(), cls)
+            execstr = "{0} = module.{0}".format(cls.__name__)
+            exec(execstr, globals(), locals())
+
+        try:
+            k = eval(kernel)
+        except:
+            raise RuntimeError("Failed to evaluate kernel string {0!r}".format(kernel))
```

A couple things about this.  
First, I've learned that it's unhelpful to just swallow exceptions that get raised.  At the very least, you should let the user know what the original exception was.
Second, if the exception is a KeyboardInterrupt, this isn't what you want.  So it is good practice whenever you catch all exceptions to first catch KeyboardInterrupt and reraise.  Apparently SystemExit is another one you should raise, but I don't know that I've ever seen a use of that one.  (cf. http://effbot.org/zone/stupid-exceptions-keyboardinterrupt.htm)

So this should become:
```
try:
    k = eval(kernel)
except (KeyboardInterrupt, SystemExit):
    raise
except Exception as e:
    raise RuntimeError("Failed to evaluate kernel string {0!r}.  Original exception: {1}".format(kernel, e))
```
Oh, and this isn't the only place where this pattern exists in Piff, so if you don't mind copying over this pattern to the other places where we were sloppy, that would be much appreciated.

---

I implemented this within gp_interp.py and gp_interp_2pcf.py.